### PR TITLE
Add support for Java 21

### PR DESCRIPTION
**Pull request by vcolin7**
This PR adds support for Java 21 runs in our CI, live tests, smoke tests, code quality reports, and aggregate reports pipelines for all our packages.

**NOTE:** Some Cosmos runs have been excluded for the time being due to their Spark connectors not being compatible with anything newer than Java 17 and Cosmos having numerous customizations for their CI and test matrices. I'll work with the Cosmos team to exclude the Spark connectors from relevant CI runs. /cc @joshfree, @kushagraThapar, @FabianMeiswinkel.

We also excluded Javadoc generation until all libraries have been migrated to use Java 21-compatible HTML.

### CODE REVIEW

**Comment by vcolin7** [comment]
Not sure why it says there ar API changes, I didn't change any public APIs in any library. Maybe it's an APIView bug @chidozieononiwu, @praveenkuttappan?

File: eng/pipelines/templates/jobs/cosmos-ci.yml
Comment (srnagar) [inline eng/pipelines/templates/jobs/cosmos-ci.yml:1]: We shouldn't ideally have custom ci jobs in `eng/pipelines` for just Cosmos alone.

File: eng/pipelines/templates/variables/globals.yml
Comment (srnagar) [inline eng/pipelines/templates/variables/globals.yml:14]: This should still be 1.20 as that is the latest non-LTS version.

File: sdk/applicationinsights/microsoft-azure-applicationinsights-query/samples/pom.xml
Comment (srnagar) [inline sdk/applicationinsights/microsoft-azure-applicationinsights-query/samples/pom.xml:20]: Do we need this for `samples/pom.xml`?

File: sdk/keyvault/platform-matrix.json
Comment (srnagar) [inline sdk/keyvault/platform-matrix.json:31]: Is this intentional to drop the version from 11 to 8?

File: sdk/loganalytics/microsoft-azure-loganalytics/pom.xml
Comment (srnagar) [inline sdk/loganalytics/microsoft-azure-loganalytics/pom.xml:126]: Any reason for removing the source and target versions? Wouldn't this change the baseline to Java 8?

File: sdk/mediaservices/microsoft-azure-media/pom.xml
Comment (srnagar) [inline sdk/mediaservices/microsoft-azure-media/pom.xml:181]: Same here - what version will the source and target be if this is removed?

File: sdk/search/azure-search-documents/pom.xml
Comment (srnagar) [inline sdk/search/azure-search-documents/pom.xml:27]: Why does the coverage drop when building with Java 21?

File: eng/versioning/version_data.txt
Comment (srnagar) [inline eng/versioning/version_data.txt:8]: We generally don't include samples here.

File: eng/pipelines/templates/jobs/cosmos-ci.yml
Comment (alzimmermsft) [reply eng/pipelines/templates/jobs/cosmos-ci.yml:1]: I'll second this. If the purpose of this is so Cosmos can pass different Java versions for build we should work on making the standard CI job more configurable. These additional CI jobs add more maintenance and potential for skew.

File: sdk/storage/azure-storage-common/pom.xml
Comment (alzimmermsft) [inline sdk/storage/azure-storage-common/pom.xml:128]: Could this be just an update to spock-core instead of adding the new dependency?

File: sdk/storage/microsoft-azure-storage-blob/pom.xml
Comment (alzimmermsft) [inline sdk/storage/microsoft-azure-storage-blob/pom.xml:99]: Same comment about being a spock-core upgrade

File: sdk/storage/platform-matrix.json
Comment (alzimmermsft) [inline sdk/storage/platform-matrix.json:17]: I would revert this back to Java 11

File: sdk/storage/platform-matrix.json
Comment (alzimmermsft) [inline sdk/storage/platform-matrix.json:23]: I'd revert this back to Java 17

**Comment by weidongxu-microsoft** [comment]
There seems change to mgmt as well. Mgmt would be fine if CI of one of them can pass.

Let me know when we plan to do the switch. Since mgmt pom.xml is generated, we would need to add the byte-buddy to codegen as default test dependency.
https://github.com/Azure/autorest.java/issues/2483

File: eng/pipelines/templates/jobs/cosmos-ci.yml
Comment (vcolin7) [reply eng/pipelines/templates/jobs/cosmos-ci.yml:1]: I agree we should not create specific templates for a few services, this would only be in place until we get the Cosmos team to add a Java 21 pipeline excluding the Spark connectors. ANother thing I can do for now is move the contents of this into the Cosmos `ci.yml` instead of having it in the `eng/` folder as a template.

File: eng/pipelines/templates/variables/globals.yml
Comment (vcolin7) [reply eng/pipelines/templates/variables/globals.yml:14]: I was thinking we could even get rid of these runs if we're doing everything but Javadoc generation on Java 21. Do we need to really run CI or live tests on a non-LTS version if it's not the latest version of Java?

File: sdk/applicationinsights/microsoft-azure-applicationinsights-query/samples/pom.xml
Comment (vcolin7) [reply sdk/applicationinsights/microsoft-azure-applicationinsights-query/samples/pom.xml:20]: When I started this work I found that if I didn't include the parent build for these samples would fail. I can go back and check if they can be removed.

File: eng/pipelines/templates/jobs/cosmos-ci.yml
Comment (vcolin7) [reply eng/pipelines/templates/jobs/cosmos-ci.yml:1]: I also created [this issue](https://github.com/Azure/azure-sdk-for-java/issues/38266) to track excluding Cosmos Spark connectors from their runs.

File: sdk/keyvault/platform-matrix.json
Comment (vcolin7) [reply sdk/keyvault/platform-matrix.json:31]: Yes. I own this service and wanted to test on Managed HSM (a more secure version of Key Vault) on the lowest and highest supported versions of Java. Now that I think about it, should we do it on every version?

File: sdk/loganalytics/microsoft-azure-loganalytics/pom.xml
Comment (vcolin7) [reply sdk/loganalytics/microsoft-azure-loganalytics/pom.xml:126]: Yes, it would change the baseline. We can keep this if you think we should.

File: sdk/search/azure-search-documents/pom.xml
Comment (vcolin7) [reply sdk/search/azure-search-documents/pom.xml:27]: I think this change came from merging `main` into this branch. I don't remember changing this manually. I'll make sure I'm not changing it compared to `main`.

File: eng/versioning/version_data.txt
Comment (vcolin7) [reply eng/versioning/version_data.txt:8]: If I didn't add them the code analysis pipeline would complain about these POMs. Maybe it's a side effect of having the samples inherit from the Azure SDK Parent? Let me see if I can remove them without that happening.

File: eng/pipelines/templates/jobs/ci.yml
Comment (vcolin7) [reply eng/pipelines/templates/jobs/ci.yml:172]: For the second point, do you mean making sure we allow libraries to generate their Javadocs on Java 21 if they already support it and then making sure we don't break things once we support Java 21 everywhere?

Comment (vcolin7) [reply eng/pipelines/templates/jobs/ci.yml:172]: Maybe you're talking about letting libraries that will never support Java 21 (Track 1, for example) Javadoc generation override this value?

File: eng/versioning/external_dependencies.txt
Comment (vcolin7) [reply eng/versioning/external_dependencies.txt:85]: We needed to add this because [the latest version](https://mvnrepository.com/artifact/org.spockframework/spock-core/2.4-M1-groovy-4.0) of Spock Core uses Groovy 4.0.6 :(

File: sdk/applicationinsights/microsoft-azure-applicationinsights-query/pom.xml
Comment (vcolin7) [reply sdk/applicationinsights/microsoft-azure-applicationinsights-query/pom.xml:131]: Yeah, I was not aware this is what caused the issues I found when trying to build these libraries using Java 21. I will look into how to build these using an earlier version of Java instead (maybe Java 17?). What do you think about that?

File: sdk/core/azure-core-experimental/src/test/java/com/azure/core/experimental/http/SimpleXmlSerializable.java
Comment (vcolin7) [reply sdk/core/azure-core-experimental/src/test/java/com/azure/core/experimental/http/SimpleXmlSerializable.java:37]: The compiler complained it was not overriding a method. I can double check.

File: sdk/storage/azure-storage-common/pom.xml
Comment (vcolin7) [reply sdk/storage/azure-storage-common/pom.xml:128]: No. We needed to add this because [the latest version](https://mvnrepository.com/artifact/org.spockframework/spock-core/2.4-M1-groovy-4.0) of Spock Core uses Groovy `4.0.6` :(

File: sdk/storage/microsoft-azure-storage-blob/pom.xml
Comment (vcolin7) [reply sdk/storage/microsoft-azure-storage-blob/pom.xml:99]: Check my earlier answer about this.

File: sdk/storage/platform-matrix.json
Comment (vcolin7) [reply sdk/storage/platform-matrix.json:17]: I see no issue with that, I just wonder why. Is it to keep things stable?

File: eng/pipelines/templates/stages/platform-matrix.json
Comment (vcolin7) [reply eng/pipelines/templates/stages/platform-matrix.json:78]: Oh right, let's remove the duplicate run. That said, I'm a little unsure about if we even need to run tests using the latest non-LTS version of Java. I get why we did it when Java 20 was the latest version of the language, but do we even need to do that now?

Comment (alzimmermsft) [reply eng/pipelines/templates/stages/platform-matrix.json:78]: The idea was, while the latest JDK available isn't LTS add in a custom run for the latest JDK to have speculative checks for things that will break once the next LTS is released. For example, if that was added with more time before Java 21 was released, it would have given us time to fix all Javadocs and once Java 21 was released we'd just flip a switch over to it, rather than what we need to do right now which is fixing everything before we can flip that switch.

And once Java 22 is released we can begin the process again. The reason it is something custom is the CI agents we use only provide LTS JDKs, everything else is something we have to install on the fly.

File: sdk/applicationinsights/microsoft-azure-applicationinsights-query/pom.xml
Comment (alzimmermsft) [reply sdk/applicationinsights/microsoft-azure-applicationinsights-query/pom.xml:131]: For Track 1 we should likely cap support at Java 8 and 11. There are some well-known issues with the design of those libraries where they attempt to do reflective accesses that break past Java 14/15, when the JDK internals were finally explicitly locked down.

File: sdk/storage/platform-matrix.json
Comment (alzimmermsft) [reply sdk/storage/platform-matrix.json:17]: I don't have a strong reason, but I also don't know why this was using Java 11 and whether that was important. Should follow-up with Storage on this

File: eng/pipelines/templates/variables/globals.yml
Comment (srnagar) [reply eng/pipelines/templates/variables/globals.yml:14]: Yes, we should support non-LTS versions to test out new features that come out in non-LTS versions. For e.g. when Java 22 rolls out, we should be able to run tests with 22 and not wait for next LTS.

File: sdk/loganalytics/microsoft-azure-loganalytics/pom.xml
Comment (srnagar) [reply sdk/loganalytics/microsoft-azure-loganalytics/pom.xml:126]: Yes, these are track 1 libraries and we should keep the baseline the same (i.e. Java 7 in this case)

File: sdk/mediaservices/microsoft-azure-media/pom.xml
Comment (srnagar) [reply sdk/mediaservices/microsoft-azure-media/pom.xml:181]: Changing the baseline will break customers if they are still on Java 7. We should not change the baseline for track 1. We either deprecate when an equivalent track 2 lib is available or continue to support existing track 1 libraries.

File: eng/pipelines/aggregate-reports.yml
Comment (joshfree) [inline eng/pipelines/aggregate-reports.yml:97]: can the comment be changed to a github issue which contains the more verbose changes required?

File: eng/pipelines/aggregate-reports.yml
Comment (joshfree) [inline eng/pipelines/aggregate-reports.yml:40]: should also reference bug number

File: eng/pipelines/templates/steps/run-and-validate-linting.yml
Comment (joshfree) [inline eng/pipelines/templates/steps/run-and-validate-linting.yml:93]: same feedback here - a bug number rather than just a TODO comment

**Review by joshfree — APPROVED** [review]
Approved, assuming feedback is addressed

File: eng/pipelines/aggregate-reports.yml
Comment (vcolin7) [reply eng/pipelines/aggregate-reports.yml:97]: Sure thing! Here you go: https://github.com/Azure/azure-sdk-for-java/issues/38267

File: sdk/core/azure-core-experimental/src/test/java/com/azure/core/experimental/http/SimpleXmlSerializable.java
Comment (vcolin7) [reply sdk/core/azure-core-experimental/src/test/java/com/azure/core/experimental/http/SimpleXmlSerializable.java:37]: Added it back.

File: sdk/search/azure-search-documents/pom.xml
Comment (vcolin7) [reply sdk/search/azure-search-documents/pom.xml:27]: Reverted this.

File: eng/versioning/version_data.txt
Comment (vcolin7) [reply eng/versioning/version_data.txt:8]: I'll keep the samples here. We do have sample Track 2 projects in the `version_client.txt` file: `azure-aot-graalvm-samples`, `azure-core-tracing-opentelemetry-samples`, `azure-identity-broker-samples`, and `azure-resourcemanager-samples`.

File: sdk/applicationinsights/microsoft-azure-applicationinsights-query/samples/pom.xml
Comment (vcolin7) [reply sdk/applicationinsights/microsoft-azure-applicationinsights-query/samples/pom.xml:20]: Other sample packages (especially Track 2) inherit from the Parent POM. I don't think it hurts to align these as well.

File: sdk/eventgrid/microsoft-azure-eventgrid/pom.xml
Comment (vcolin7) [reply sdk/eventgrid/microsoft-azure-eventgrid/pom.xml:100]: Reverted this change.

File: sdk/batch/microsoft-azure-batch/pom.xml
Comment (vcolin7) [reply sdk/batch/microsoft-azure-batch/pom.xml:162]: Reverted this change.

File: sdk/applicationinsights/microsoft-azure-applicationinsights-query/pom.xml
Comment (vcolin7) [reply sdk/applicationinsights/microsoft-azure-applicationinsights-query/pom.xml:131]: Reverted this change. I noticed we build Track 1 libraries using Java 11 in the template YAML.

File: sdk/mediaservices/microsoft-azure-media/pom.xml
Comment (vcolin7) [reply sdk/mediaservices/microsoft-azure-media/pom.xml:181]: Reverted this change.

**Comment by vcolin7** [comment]
> There seems change to mgmt as well. Mgmt would be fine if CI of one of them can pass.
> 
> Let me know when we plan to do the switch. Since mgmt pom.xml is generated, we would need to add the byte-buddy to codegen as default test dependency. [Azure/autorest.java#2483](https://github.com/Azure/autorest.java/issues/2483)

Hi @weidongxu-microsoft, we'll likely do this switch early next week (Jan 15th - Jan 19th) as this PR is ready to merge. Let me know if you need to make the management POM change first. I'll hold off from merging from the time being :)

**Comment by weidongxu-microsoft** [comment]
> Hi @weidongxu-microsoft, we'll likely do this switch early next week (Jan 15th - Jan 19th) as this PR is ready to merge. Let me know if you need to make the management POM change first. I'll hold off from merging from the time being :)

@vcolin7

Please go ahead with your schedule.

The code in codegen is already merged.
https://github.com/Azure/autorest.java/pull/2490
We should release a new version of autorest-java with the code (after azure-core release of Jan. and before we release any new mgmt package).

**Comment by weidongxu-microsoft** [comment]
@vcolin7

One question though: why we use prefix `testdep_` on these?
```
net.bytebuddy:byte-buddy;1.12.23
testdep_net.bytebuddy:byte-buddy;1.14.8
testdep_net.bytebuddy:byte-buddy-agent;1.14.8
```

**Comment by vcolin7** [comment]
> @vcolin7
> 
> One question though: why we use prefix `testdep_` on these?
> 
> ```
> net.bytebuddy:byte-buddy;1.12.23
> testdep_net.bytebuddy:byte-buddy;1.14.8
> testdep_net.bytebuddy:byte-buddy-agent;1.14.8
> ```

I think @alzimmermsft introduced the prefix to pin this specific version that Mockito `4.11` needs and to avoid overriding it in any other place where ByteBuddy needs to be on a more recent version.

**Comment by alzimmermsft** [comment]
> > @vcolin7
> > One question though: why we use prefix `testdep_` on these?
> > ```
> > net.bytebuddy:byte-buddy;1.12.23
> > testdep_net.bytebuddy:byte-buddy;1.14.8
> > testdep_net.bytebuddy:byte-buddy-agent;1.14.8
> > ```
> 
> I think @alzimmermsft introduced the prefix to pin this specific version that Mockito `4.11.0` needs and to avoid overriding it in any other place where ByteBuddy needs to be on a more recent version.

That and I believe there is some tooling that could overwrite this value if it was just `net.bytebuddy...` and I wanted to make it explicitly clear this was a testing dependency only.

**Comment by Netyyyy** [comment]
Hi @vcolin7 , `JavadocSafeJavaBuildVersion` doesn't work unless adding `JavadocSafeJavaBuildVersion: ${{ parameters.JavadocSafeJavaBuildVersion }} ` into parameters: https://github.com/Azure/azure-sdk-for-java/blob/2adb0f45ca7f1925a9a6fbb9e80d5c1a5b2ae38f/eng/pipelines/templates/stages/archetype-sdk-client.yml#L81-L85

**Pull request merged**
